libflux/src/core/formatter/mod.rs

```diff
@@ -41,28 +50,76 @@ pub struct Formatter {
     //     1
     temp_indent: bool,
     err: Option<Error>,
+
+    // temp_tabstops indicates if we should be inserting tabstops when formatting nodes.
+    temp_tabstops: bool,
 }
 
+// INDENT_BYTES is 4 spaces as a constant byte slice
+const INDENT_BYTES: [u8; 4] = [32, 32, 32, 32];
+
 impl Default for Formatter {
     fn default() -> Self {
         Formatter {
-            builder: String::new(),
+            // The TabWriter replaces \t characters with spaces according to its vertical alignment
+            // rules.
+            // We set the padding and minwidth to 0 so that \t characters are removed unless they
+            // appear in a set of adjacent lines in the output source.
+            builder: TabWriter::new(vec![])
+                .alignment(Alignment::Right)
+                .padding(0)
+                .minwidth(0),
             indentation: 0,
             clear: true,
             temp_indent: false,
             err: None,
+            temp_tabstops: false,
+        }
+    }
+}
+
+//
+// Implement specific error From conversions based on the kinds of errors we can encounter
+//
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Self {
+        Error {
+            msg: format!("{}", err),
+        }
+    }
+}
+impl From<FromUtf8Error> for Error {
+    fn from(err: FromUtf8Error) -> Self {
+        Error {
+            msg: format!("{}", err),
+        }
+    }
+}
+
+impl From<IntoInnerError<TabWriter<Vec<u8>>>> for Error {
+    fn from(err: IntoInnerError<TabWriter<Vec<u8>>>) -> Self {
+        Error {
+            msg: format!("{}", err),
         }
     }
 }
 
 impl Formatter {
     // returns the final string and error msg
-    pub fn output(&self) -> Result<String, String> {
-        if let Some(err) = &self.err {
-            return Err(err.msg.clone());
+    pub fn output(mut self) -> Result<String, Error> {
+        if let Some(err) = self.err {
+            return Err(err);
         }
 
-        Ok(self.builder.clone())
+        self.builder.flush()?;
+        Ok(String::from_utf8(self.builder.into_inner()?)?)
```

Comment:
While this is a Correct Rust Idiom™, the use of `into_inner` to convert from type to type has always been a little awkward to me. Even acknowledging that rust does it "right" with requiring implementations of `From`, it still feels duck-typey in a way that feels wrong in rust.

Nothing actionable here. I'm just talking out loud.

Reply:
yeah, this is the _only_ way to do it according to the TabWriter API, but yeah its a bit round about.